### PR TITLE
Reduce allocations through a few tricks

### PR DESCRIPTION
**Pull request by goldfirere**
Guided by memtrace, I identified spots with lots of allocations and fixed them, all in separate commits. Taken together, these commits lead to a 2.8% reduction in allocations. (That's quite different than what you get if you sum up the individual commits, but it's what I got when I remeasured at the end.)

Note that one of these commits introduces `[@local]` into the compiler source. I think that's OK. But maybe you disagree.

### CODE REVIEW

**Comment by antalsz** [comment]
Before I review, I'm concerned about `[@local]`: As @stedolan put it, we want to be able to compile the compiler with "any `ocaml` we find by the side of the road".  While the attribute form doesn't violate that, we're planning to at a minimum make the attribute form much uglier as I refactor everything to use the modular extensions machinery, and rely on ocamlformat for upstreamability.  This feels like it might cause problems as a result.

File: ocaml/typing/subst.ml [hunk 1]
```diff
@@ -188,118 +188,123 @@ let norm = function
...
-      | Tobject (t1, name) ->
-          let t1' = typexp copy_scope s t1 in
-          let name' =
-            match !name with
-            | None -> None
-            | Some (p, tl) ->
-                if to_subst_by_type_function s p
-                then None
-                else Some (type_path s p, List.map (typexp copy_scope s) tl)
+let typexp copy_scope s =
```
Comment (antalsz) [inline ocaml/typing/subst.ml:191]: For future reviewers: this whole change is much more easily reviewable with `patdiff`, it's just a change from `typeexp copy_scope s` to `loop` everywhere below

**Review by antalsz — CHANGES_REQUESTED** [review]
I'm not an efficiency expert, but I confirm that these changes look semantically idempotent!  It's interesting that the change in `subst` is sort of the opposite of the usual change.

My one remaining concern is around `[@local]`, as I said above.  Other that that, this looks good to me, although I haven't rerun this to get my own performance numbers.

**Comment by goldfirere** [comment]
> Before I review, I'm concerned about `[@local]`: As @stedolan put it, we want to be able to compile the compiler with "any `ocaml` we find by the side of the road". While the attribute form doesn't violate that, we're planning to at a minimum make the attribute form much uglier as I refactor everything to use the modular extensions machinery, and rely on ocamlformat for upstreamability. This feels like it might cause problems as a result. (And we certainly can't upstream it.)

I'm also worried about this, which is why I highlighted the problem. There are other places that I noticed that would require `[%local]`, which is definitely worse. (And actually didn't work in the dune-based build, but that's a different story.)

Maybe we've gotten the plan slightly wrong around extensions. We had previously thought that erasable extensions (ones that could be removed uniformly without changing correctness) would be attributes and non-erasable ones would be extension nodes. With our new plan for dealing with public-release using a translation tool (involving ocamlformat), we realized we didn't need to distinguish the syntax of the two. But perhaps we still should, but a different way: all extensions are attributes, but erasable ones have a short form. Considering we're hoping to attract a wider audience to our version of ocaml via opam support, other users might want to put erasable extensions in their code, too. The short form facilitates this. And then we're on more solid ground putting them into the compiler itself.

**Comment by goldfirere** [comment]
Should we merge this then?

**Review by antalsz — APPROVED** [review]
I think I'm convinced by your argument for attributes, which will affect how I'm designing the revamp of the extensions machinery.  It would still be good to get an efficiency-minded review of the code, but I rescind my objections

**Comment by goldfirere** [comment]
So this has now run through the internal perf testing process, under test ID 3338. I'm not the best poised to interpret the results -- there seems to be a modest improvement, though quite a bit less than I was expecting.

**Comment by goldfirere** [comment]
@ccasin I'd love your knowledgeable opinion on the perf results. I don't quite know how to interpret them.

**Comment by ccasin** [comment]
I'm clearly the wrong person to evaluate this kind of thing, but, looking quickly:
- Looks like this saves about 1% of overall allocations (comparing your 3338 with recent builds of the same profile, like 3294 and 3343).  Nice!  Worth having.
- I'd love to hear from @mshinwell or someone else knowledgeable about our optimizers regarding whether we should be expecting the compiler to perform transformations like this for us, and about what might account for the large gap between the difference you saw locally and the one we see on the full build.
- I'd love to hear from @lpw25 about whether there are reasons not to put `@local` in the compiler.  If we're going to allow for our extensions in our copy of the compiler, there are a lot more opportunities here!

**Comment by lpw25** [comment]
I don't mind using `[@local]` in the tree since it will still build with other compilers.

It's probably worth saying that I think your mental model for performance is a little off though. In particular, you should be careful using total allocations as a proxy for performance.

Total allocations can be a very useful measure because idiomatic OCaml code allocates frequently, so total allocations can be a reasonable approximation for total time -- a sudden increase in allocations probably corresponds to the program doing a lot more work. It can also give you a sense for how compact the memory representations people are using are. So reductions in total allocations that are achieved by doing less work or by using a more compact memory representation will probably correspond to improvements in performance.

Short-lived allocations themselves though are very cheap and mostly don't contribute much to performance. There are four ways that they might contribute:

1. All allocations advance "time" for the GC. That means that the more allocations you do the faster other allocations age, meaning that more allocations will be long-lived and therefore costly. However, this is best measured directly by looking at major heap allocations rather than total allocations, or even more directly by looking at how much time is spent marking and sweeping in a perf profile.
2. Similarly, there is some fixed cost for performing minor GCs and they happen after a fixed amount of "time" has passed. Again, you can see this one directly by looking at total number of minor GCs, or by looking at how much time is spent doing minor gc work in a perf profile.
3. If your minor heap doesn't fit entirely in cache then by doing an allocation your active data now uses more cache lines that it did before.
4. There is a very small amount of work to do the allocation itself. I wouldn't expect this to show up outside of extremely tight loops though.

Making a short-lived allocation local always removes cost 1 and 2, and may remove cost 3 depending on the details of the code by allowing you to reuse the same cache line for multiple allocations. It doesn't affect 4. All four gains tend to be much smaller than the gains from doing less underlying work or using more compact memory representations. They matter for very hot code, but I think I'd be surprised if a 3% drop in allocations achieved by removing unnecessary allocations improved performance by enough to distinguish it from noise.

That's not to say I'm against the changes in this patch. Far from it: I kind of hate pointless closure allocations and will almost always remove them when I notice them. Even if they don't cost much, they cost something, and they make your allocation profiles less useful by hiding the important long-lived data. I just don't think you want to spend too much time optimising the code in this way because you aren't likely to see large improvements from it. At the very least you should be looking at measures beyond total allocations to assess the value of the changes.

**Comment by goldfirere** [comment]
@lpw25 

> It's probably worth saying that I think your mental model for performance is a little off though. In particular, you should be careful using total allocations as a proxy for performance.

Duly noted. The next time I do this sort of work, I will look at these other measures (too). Thanks for the explanation.

**Comment by mshinwell** [comment]
@goldfirere is this to be merged?  (needs rebase)

**Comment by goldfirere** [comment]
I don't quite know what to do with this PR. It makes the code worse. Yet it really did reduce allocations. @lpw25 comments above that I might have been measuring the wrong thing. In our internal benchmarks, this doesn't rise above the noise threshold, so it's hard to know if there's a real improvement.

Maybe the best thing is to revisit this when/if we have more accurate benchmarks? I suppose that, until then, a change that falls below the noise threshold is effectively not a change, and so we should abandon. I'd love others' feedback on this.

**Comment by goldfirere** [comment]
Closing after some discussion. These might be good ideas, but they're better evaluated once we are using Flambda2 classic mode for fast-build.

**Pull request closed without merging**